File: Data Structures & Algorithms/subsets-ii/submission-4.java
class Solution {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums); // Step 1: sort to group duplicates
        List<List<Integer>> global = new ArrayList<>();
        global.add(new ArrayList<Integer>());
        subsetsH(nums, 0, global, new ArrayList<Integer>());
        return global;
    }

    public void subsetsH(
        int[] nums, 
        int i, 
        List<List<Integer>> global, 
        List<Integer> res
    ) {
        for (int j = i; j < nums.length; j++) {
            // Step 2: skip duplicates
            if (j > i && nums[j] == nums[j - 1]) continue;

            res.add(nums[j]);
            global.add(new ArrayList<>(res));
            subsetsH(nums, j + 1, global, res);
            res.remove(res.size() - 1); // backtrack
        }
    }
}

File: Data Structures & Algorithms/two-integer-sum-ii/submission-4.py
class Solution:
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        
        numbers.sort()
        res = []
        l: int = 0
        r: int = len(numbers) - 1    
        
        while l < r:
            sum: int = numbers[l] + numbers[r]
            if sum == target:
                res = [l+1, r+1]
                l += 1
                r -= 1
            elif sum < target:
                l += 1
            else:
                r -= 1

        res.sort()     
        return res         



             
        